res/controllers/Pioneer-DDJ-SR2-scripts.js

```diff
@@ -0,0 +1,1459 @@
+    0xFF00FF: DDJSR2.wheelLedCircleColor.MAGENTA,
+    0xEE82EE: DDJSR2.wheelLedCircleColor.VIOLET,
+    0x8800CC: DDJSR2.wheelLedCircleColor.PURPLE,
+    0xFFFFFF: DDJSR2.wheelLedCircleColor.WHITE,
+});
+
+
+////////////
+// Code. //
+//////////
+
+DDJSR2.seratoHeartbeat=[0xF0, 0x00, 0x20, 0x7F, 0x50, 0x01, 0xF7]; // Serato Heartbeat
+DDJSR2.seratoControlsPoll=[0xF0, 0x00, 0x20, 0x7F, 0x03, 0x01, 0xF7]; // Serato first poll
+
+DDJSR2.init = function() {
+    // Browser section is part of Mixer container for now.
+    
+    DDJSR2.browser = new DDJSR2.BrowserContainer();
+
+    DDJSR2.deck = [];
+    for (var i = 0; i < 4; i++) {
+        DDJSR2.deck[i] = new DDJSR2.Deck(i);
+        DDJSR2.deck[i].setCurrentDeck("[Channel" + (i + 1) + "]");
+    }
+
+    DDJSR2.mixer = new DDJSR2.MixerContainer();
+
+    DDJSR2.createEffectUnits();
+
+    // Performance pads are part of the Deck specs, as are Illumination Controls
+
+    DDJSR2.seratoHeartbeatTimer = engine.beginTimer(250, DDJSR2.doSeratoHeartbeatTimer, false);
+
+    // After 500 ms have elapsed, after the serato heartbeat has occurred, then poll controls. Calling this before time results in an unsuccessful poll.
+
+ engine.beginTimer(500, function(){
+	midi.sendSysexMsg(DDJSR2.seratoControlsPoll, DDJSR2.seratoControlsPoll.length)}
+	,true);
+
+	  engine.beginTimer(600, function(){ midi.sendShortMsg(0x9B,0x00,0x00) }
+        ,true);
+
+};
+
+DDJSR2.doSeratoHeartbeatTimer = function() {
+    midi.sendSysexMsg(DDJSR2.seratoHeartbeat, DDJSR2.seratoHeartbeat.length);
+};
+
+
+// First, a function library.
+////////////////
+
+
+DDJSR2.getRotaryDelta = function(value) {
+    let delta = 0x40 - Math.abs(0x40 - value),
+        isCounterClockwise = value > 0x40;
+
+    if (isCounterClockwise) {
+        delta *= -1;
+    }
+
+    return delta;
+};
+
+DDJSR2.getJogWheelDelta = function(value) {
+    // The Wheel control centers on 0x40; find out how much it's moved by.
+    return value - 0x40;
+};
+
+DDJSR2.wheelLedControl = function(channelOffset, color) {
+    const wheelLedBaseChannel = 0xBB;
+    if (channelOffset !== null) {
+        midi.sendShortMsg(
+            wheelLedBaseChannel,
+            0x20+(channelOffset),
+            color 
+        );
+    }
+};
+
+
+DDJSR2.pitchBendFromJog = function(group, movement) {
+    engine.setValue(group, "jog", movement / 5 * DDJSR2.jogwheelSensitivity);
+};
+
+DDJSR2.ringBufferView = class {
+    constructor(indexable, startIndex = 0) {
+        this.indexable = indexable;
+        this.index = startIndex;
+    }
+    advanceBy(n) {
+        this.index = script.posMod(this.index + n, this.indexable.length);
+        return this.current();
+    }
+    next() {
+        return this.advanceBy(1);
+    }
+    previous() {
+        return this.advanceBy(-1);
+    }
+	current() {
+        return this.indexable[this.index];
+    }
+};
+
+DDJSR2.BrowserContainer = function(){
+
+    this.browseKnob  = new components.ComponentContainer({
```

Comment:
please fix the formatting here (pre-commit should probably take care of it). One level of indentation per opening brace.